utilities/cassandra/cassandra_compaction_filter.cc

```diff
@@ -32,6 +32,10 @@ CompactionFilter::Decision CassandraCompactionFilter::FilterV2(
           ? row_value.RemoveExpiredColumns(&value_changed)
           : row_value.ConvertExpiredColumnsToTombstones(&value_changed);
 
+  if (value_type == ValueType::kValue) {
```

Comment:
> This PR also removes an old optimization in cassandra merge operator for single merge operands. We need to do GC even on a single operand, so the optimation does not make sense anymore.

 I'm a fan of consolidating purge logic into compaction filter. Should we call RemoveTombstone on all value_type instead of kValue?

Reply:
I guess I too agree that it is much cleaner to call RemoveTombstones() irrespective or the value type. 
Another suggestion I have here is that, now you will be calling `RemoveExpiredColumns` / `ConvertExpiredColumsToTombstones` followed by `RemoveTombstones` for every kv pair. All these functions have the common code of column-iterations-loop. So you end up doing double the work than what you really need. May be you could come up with a single function call instead of two between lines 30 - 37 ?

Reply:
@sagar0 @scv119  in compaction filter removing tombstone for a merge type is not safe. If we remove it there we run the risk of old data reappearing even gc grace passed. 

That said I do feel a pain splitting the GC logic between merge operator and compaction filter. But we can consolidate it into one place only if we have a new compaction filter interface trigger *after* merge is done. @sagar0, will it need a lot of efforts for such an interface?

@sagar0 Regarding your suggestion: given the behavior is dynamic, if we have one function we will need passing in two boolean states (whether should remove or convert expired column, whether should collect tombstone). Feels not very clean. And probably there is not much performance gain from less looping. Because in production the number of columns is normally only 1. For the common code between those three functions, I can do some refactoring to reduce the duplication.

Reply:
@wpc thanks for explaining. your reasoning sounds good to me. I am fine with keeping the call to RemoveTombstones under the kValue type conditional.

Re. the order of compaction filter and merge operator calls during compactions:  The current order has been in use, I believe, since quite some time. I am not sure of the impact it might have on the other users if that order is changed. However, we can definitely look into modifying the order if there is really a perf impact that we are seeing. I suggest we continue with the current order for V1, and evaluate its performance.